### PR TITLE
feat(Boost for Reddit): Add `Fix missing audio in video downloads` patch

### PR DESCRIPTION
**Pull request by OctoNezd**
Reddit changed their audio endpoints around the time API controversy started: https://old.reddit.com/r/redditdev/comments/15cdln3/reddit_changed_what_url_audio_from_a_video_is/

Boost has proper audio playback in video player itself - I assume it parses DASHPlaylist.mpd, but not when downloading - downloads have these hardcoded paths: `/dash_audio.mp4` and if that fails then it tries to use `/audio.mp4`. This PR replaces these values with `/DASH_AUDIO_128.mp4` and `/DASH_AUDIO_64.mp4`

### CODE REVIEW

File: src/main/kotlin/app/revanced/patches/reddit/customclients/boostforreddit/fix/downloadVReddItAudio/DownloadVReddItAudio.kt [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+import app.revanced.patcher.data.BytecodeContext
+import app.revanced.patcher.extensions.InstructionExtensions.replaceInstructions
+import app.revanced.patcher.patch.BytecodePatch
+import app.revanced.patcher.patch.annotation.CompatiblePackage
+import app.revanced.patcher.patch.annotation.Patch
+import app.revanced.patches.reddit.customclients.boostforreddit.fix.downloadVReddItAudio.fingerprints.DownloaderSelectVReddItAudioLink
+
+@Patch(
+    name="Fix audio in videos downloaded from v.redd.it",
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/reddit/customclients/boostforreddit/fix/downloadVReddItAudio/DownloadVReddItAudio.kt:11]: What exactly does this fix? Downloading videos, just the audio?

File: src/main/kotlin/app/revanced/patches/reddit/customclients/boostforreddit/fix/downloadVReddItAudio/DownloadVReddItAudio.kt [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    compatiblePackages = [CompatiblePackage("com.rubenmayayo.reddit")]
+)
+object FixDownloadedVReddItAudio : BytecodePatch(setOf(DownloaderSelectVReddItAudioLink)) {
+    override fun execute(context: BytecodeContext) {
+        DownloaderSelectVReddItAudioLink.result?.mutableMethod?.replaceInstructions(17, "const-string v3, \"/DASH_AUDIO_128.mp4\"")
+        DownloaderSelectVReddItAudioLink.result?.mutableMethod?.replaceInstructions(34, "const-string p2, \"/DASH_AUDIO_64.mp4\"")
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/reddit/customclients/boostforreddit/fix/downloadVReddItAudio/DownloadVReddItAudio.kt:17]: Please use mutableMethod?.apply to reduce duplication.
The magic constants such as the indices as well as the registers must be referenced from the original code. Use .getInstruction<Instruction>() in order to get them.

File: src/main/kotlin/app/revanced/patches/reddit/customclients/boostforreddit/fix/downloadVReddItAudio/DownloadVReddItAudio.kt (hunk 1 above)
Comment (OctoNezd) [reply src/main/kotlin/app/revanced/patches/reddit/customclients/boostforreddit/fix/downloadVReddItAudio/DownloadVReddItAudio.kt:11]: Videos are downloaded/shared without proper audio. Playback in-app works fine. 

With this fix, downloads and shares would have audio in them as well.

File: src/main/kotlin/app/revanced/patches/reddit/customclients/boostforreddit/fix/downloadVReddItAudio/DownloadVReddItAudio.kt (hunk 2 above)
Comment (OctoNezd) [reply src/main/kotlin/app/revanced/patches/reddit/customclients/boostforreddit/fix/downloadVReddItAudio/DownloadVReddItAudio.kt:17]: How can I use getInstruction to find index? I dont see it accepting anything except indexes

Comment (oSumAtrIX) [reply src/main/kotlin/app/revanced/patches/reddit/customclients/boostforreddit/fix/downloadVReddItAudio/DownloadVReddItAudio.kt:17]: Either you use the indexOfFirstInstruction extension function or get the location from a builder instruction in a mutable method

**Comment by OctoNezd** [comment]
Is this the desired version?

File: src/main/kotlin/app/revanced/patches/reddit/customclients/boostforreddit/fix/downloadVReddItAudio/DownloadVReddItAudio.kt [hunk 3]
```diff
@@ -7,23 +7,37 @@ import app.revanced.patcher.patch.BytecodePatch
...
+
     override fun execute(context: BytecodeContext) {
-        DownloaderSelectVReddItAudioLink.result?.mutableMethod?.apply {
-            INDEXES.forEach {
-                val instruction = getInstruction(it.first) as Instruction21c
-                replaceInstruction(it.first, "const-string v${instruction.registerA}, \"${it.second}\"")
+        val searchResult = DownloaderSelectVReddItAudioLink.result
+        searchResult?.mutableMethod?.apply {
+            INDEXES.forEach { endpointPair ->
+                val instructionIndex = searchResult.scanResult.stringsScanResult!!.matches.find {
```
[the remark this answers is not in the document]
Comment (OctoNezd) [reply src/main/kotlin/app/revanced/patches/reddit/customclients/boostforreddit/fix/downloadVReddItAudio/DownloadVReddItAudio.kt]: Replaced with string index as you suggested

**Comment by OctoNezd** [comment]
Yes, it works

File: src/main/kotlin/app/revanced/patches/reddit/customclients/boostforreddit/fix/downloads/fingerprints/DownloaderSelectVReddItAudioLink.kt [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package app.revanced.patches.reddit.customclients.boostforreddit.fix.downloads.fingerprints
+
+import app.revanced.patcher.fingerprint.MethodFingerprint
+
+internal object DownloaderSelectVReddItAudioLink : MethodFingerprint(
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/reddit/customclients/boostforreddit/fix/downloads/fingerprints/DownloaderSelectVReddItAudioLink.kt:5]: Please rename this fingerprint. What does the method this fingerprint resolves to, do?

Comment (OctoNezd) [reply src/main/kotlin/app/revanced/patches/reddit/customclients/boostforreddit/fix/downloads/fingerprints/DownloaderSelectVReddItAudioLink.kt:5]: This method checks if the link is v.redd.it, and if it is, it first tries to download .../DASH_AUDIO.mp4, and if that fails then .../audio, and if that fails as well then it just gives up trying to download audio.

**Review by oSumAtrIX — APPROVED** [review]

**Comment by VaiCaDep0893** [comment]
don't forget that infinity also has this issue

**Comment by OctoNezd** [comment]
Infinity is a project that has maintainer. And it seems that this issue is fixed: https://github.com/Docile-Alligator/Infinity-For-Reddit/commit/c5243ade5647b6a788b6c0a318863caad57a18f7

**Comment by leandroprz** [comment]
This patch isn't working for me. I keep getting videos without audio.

**Comment by OctoNezd** [comment]
Can confirm this got broken by reddit again.

**Comment by OctoNezd** [comment]
Looks like app isnt even trying to download audio. I will look into it hopefully this weekend.

**Comment by OctoNezd** [comment]
Fix for that feature is in #3379

**Pull request merged**